### PR TITLE
[Enhancement] optimize or predicate mv rewrite

### LINKED ISSUE #26038: optimize mv rewrite to support or predicate rewrite
## Enhancement
for mv:

select * from lineorder where lo_orderdate < "20201001" or lo_orderdate > "20210101";

for query:
select * from lineorder where lo_orderdate > "20210201";

the predicate in query  lo_orderdate > "20210201" is subset of mv, query is expected to be rewritten by mv.
now the query can not be rewritten by mv because or predicate is not processed well.

we should support this type of rewrite.

### PR DESCRIPTION
**Pull request by ABingHuang**
Fixes #26038 

Optimize or predicate rewrite and add support to rewrite in predicate.

## What type of PR is this:
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/ColumnRangePredicate.java [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+import com.starrocks.sql.optimizer.operator.scalar.BinaryPredicateOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ColumnRefOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ConstantOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ScalarOperator;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+public class ColumnRangePredicate extends RangePredicate {
```
Comment (LiShuMing) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/ColumnRangePredicate.java:32]: Add some docs about this class?

It represents all ranges about one column ref?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/PredicateExtractor.java [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+    public RangePredicate visit(ScalarOperator scalarOperator, PredicateExtractorContext context) {
+        return null;
+    }
+
+    @Override
+    public RangePredicate visitBinaryPredicate(
+            BinaryPredicateOperator predicate, PredicateExtractorContext context) {
+        ScalarOperator left = predicate.getChild(0);
+        ScalarOperator right = predicate.getChild(1);
+        if (left.isColumnRef() && right.isConstantRef()) {
```
Comment (LiShuMing) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/PredicateExtractor.java:65]: How about right is const re f and left is column ref?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/PredicateExtractor.java [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+            residualPredicates.add(predicate);
+        }
+        return null;
+    }
+
+    @Override
+    public RangePredicate visitCompoundPredicate(
+            CompoundPredicateOperator predicate, PredicateExtractorContext context) {
+        if (predicate.isNot()) {
+            return null;
```
Comment (LiShuMing) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/PredicateExtractor.java:86]: can we support `not` later?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/PredicateExtractor.java [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+
+                } else {
+                    rangePredicates.add(childRange);
+                }
+            } else if (predicate.isNot()) {
+                return null;
+            }
+        }
+        if (predicate.isAnd()) {
+            return new AndRangePredicate(rangePredicates);
```
Comment (LiShuMing) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/PredicateExtractor.java:149]: `rangePredicates ` may contain both or  / and  predicates?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/RangePredicate.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+
+package com.starrocks.sql.optimizer.rule.transformation.materialization;
+
+import com.starrocks.sql.optimizer.operator.scalar.ScalarOperator;
+
+public class RangePredicate {
```
Comment (LiShuMing) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/RangePredicate.java:20]: Add some descriptions for each method?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/OrRangePredicate.java [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+    public OrRangePredicate(List<RangePredicate> rangePredicates) {
+        this.rangePredicates = rangePredicates;
+    }
+
+    public List<RangePredicate> getRangePredicates() {
+        return rangePredicates;
+    }
+
+    @Override
+    public boolean enclose(RangePredicate other) {
```
Comment (LiShuMing) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/OrRangePredicate.java:38]: This `OrRangePredicate ` can enclose `other` predicates?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/ColumnRangePredicate.java (hunk 1 above)
Comment (ABingHuang) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/ColumnRangePredicate.java:32]: yes

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/PredicateExtractor.java (hunk 2 above)
Comment (ABingHuang) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/PredicateExtractor.java:65]: when enter this logic, the scalar operator has been normalized, so here could only be left is column and right is constant. I will add protection logic here.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/PredicateExtractor.java (hunk 3 above)
Comment (ABingHuang) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/PredicateExtractor.java:86]: I will add this function

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/PredicateExtractor.java (hunk 4 above)
Comment (ABingHuang) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/PredicateExtractor.java:149]: yes.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/RangePredicate.java (hunk 5 above)
Comment (ABingHuang) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/RangePredicate.java:20]: ok

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/OrRangePredicate.java (hunk 6 above)
Comment (ABingHuang) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/OrRangePredicate.java:38]: yes, check whether OrRangePredicate can enclose other predicates.

**Review by murphyatwork — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 318 / 370 (85.95%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/RangePredicate.java|4|10|40.00%|[31, 32, 33, 46, 55, 63]|
|:large_blue_circle:|com/starrocks/catalog/PartitionKeyDiscreteDomain.java|3|7|42.86%|[33, 38, 42, 47]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/scalar/MvNormalizePredicateRule.java|18|28|64.29%|[133, 136, 146, 148, 151, 162, 163, 165, 167, 168]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/OrRangePredicate.java|23|32|71.88%|[37, 38, 39, 40, 41, 44, 47, 49, 96]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/AndRangePredicate.java|58|71|81.69%|[33, 34, 35, 36, 37, 39, 40, 41, 44, 45, 49, 50, 137]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/PredicateExtractor.java|118|125|94.40%|[71, 72, 73, 74, 176, 178, 226]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/ColumnRangePredicate.java|65|68|95.59%|[82, 128, 163]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/PredicateSplit.java|11|11|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MvUtils.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/RangeSimplifier.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/PartitionColPredicateEvaluator.java|8|8|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/scalar/MvNormalizePredicateRule.java [hunk 7]
```diff
@@ -123,6 +126,47 @@ public ScalarOperator visitCompoundPredicate(CompoundPredicateOperator predicate
...
+                .collect(Collectors.toList());
+        if (constants.size() == 1) {
+            BinaryType op =
+                    isIn ? BinaryType.EQ : BinaryType.NE;
+            result.add(new BinaryPredicateOperator(op, lhs, constants.get(0)));
+        } else if (!constants.isEmpty()) {
+            for (ScalarOperator constant : constants) {
+                BinaryType op =
+                        isIn ? BinaryType.EQ : BinaryType.NE;
+                result.add(new BinaryPredicateOperator(op, lhs, constant));
```
[the remark this answers is not in the document]
Comment (ABingHuang) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/scalar/MvNormalizePredicateRule.java:156]: add a limit to 1024

**Review by murphyatwork — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

**Pull request merged**